### PR TITLE
[ADAM-1978] Add additional filter by convenience methods.

### LINKED ISSUE #1978: Add additional filter by convenience methods
See comment https://github.com/bigdatagenomics/adam/issues/1877#issuecomment-358049617

### PR DESCRIPTION
**Pull request by heuermh**
Fixes #1978.

Thank you to @antonkulaga, the methods here are inspired by those in
https://github.com/antonkulaga/adam-playground

### CODE REVIEW

**Comment by heuermh** [comment]
@fnothaft Milestone as 0.24.1?

**Comment by heuermh** [comment]
Feel free to make suggestions.  There is a balance between too few and too many

VariantRDD/Dataset
```scala
filterByFiltersPassed() // VCF column 7 "FILTER"
filterByQuality(minimumQuality) // VCF column 6 "QUAL"
filterByReadDepth(minimumReadDepth)  // VCF INFO AD, Number=R split for multi-allelic
filterByReferenceReadDepth(minimumReferenceReadDepth) // VCF INFO AD, Number=R split for multi-allelic

filterSNPs()
filterMNPs()
filterIndels()

filterToSNPs()
filterToMNPs()
filterToIndels()
```

GenotypeRDD/Dataset
```scala
filterByFiltersPassed() // VCF FORMAT FT
filterByQuality(minimumQuality) // VCF FORMAT GQ
filterByReadDepth(minimumReadDepth) // VCF FORMAT DP
filterByAlternateReadDepth(minimumAlternateReadDepth) // VCF FORMAT AD, Number=R split for multi-allelic sites
filterByReferenceReadDepth(minimumReferenceReadDepth) // VCF FORMAT AD, Number=R split for multi-allelic sites

filterToSample(sampleId: String)
filterToSamples(sampleIds: Seq[String])

filterNoCalls()
filterHomozygousReference()
```

AlignmentRecordRDD/Dataset
```scala
filterByMapq(minimumMapq)

filterUnalignedReads()
filterUnpairedReads()
filterDuplicateReads()

filterToPrimaryAlignments()

filterToRecordGroup(recordGroupName: String)
filterToRecordGroups(recordGroupNames: Seq[String])

filterToSample(recordGroupSample: String)
filterToSamples(recordGroupSamples: Seq[String])
```

**Comment by heuermh** [comment]
Ping for addl feedback before I get too far along with unit tests.

**Comment by heuermh** [comment]
I tried to be careful with method names.

Methods that filter out/remove matching records are named `filterXxx`
```scala
def filterUnalignedReads(): AlignmentRecordRDD
```

Methods that filter out/remove records that do not match are named `filterToXxx(s)`.
```scala
def filterToParent(parentId: String): FeatureRDD
def filterToPrimaryAlignments(): AlignmentRecordRDD
```

Methods that filter on numeric values are named `filterByXxx` and the method parameter is consistently `minimumXxx`
```scala
def filterByScore(minimumScore: Double): FeatureRDD
```

A few don't match this pattern, perhaps they should also be `filterToXxx`
```scala
def filterByFeatureType(featureType: String): FeatureRDD
def filterByAttribute(key: String, value: String): FeatureRDD
```

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala [hunk 1]
```diff
@@ -315,6 +315,34 @@ case class DatasetBoundFeatureRDD private[rdd] (
       .withColumnRenamed("score", "count")
       .as[Coverage], sequences)
   }
+
+  override def filterByFeatureType(featureType: String): FeatureRDD = {
+    transformDataset(dataset => dataset.filter(dataset.col("featureType").eqNullSafe(featureType)))
+  }
+
+  override def filterToGene(geneId: String): FeatureRDD = {
```
[the remark this answers is not in the document]
Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala]:
```scala
def filterToGenes(geneIds: Seq[String]): FeatureRdd
```
?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala [hunk 2]
```diff
@@ -315,6 +315,34 @@ case class DatasetBoundFeatureRDD private[rdd] (
...
+
+  override def filterToTranscript(transcriptId: String): FeatureRDD = {
+    transformDataset(dataset => dataset.filter(dataset.col("transcriptId").eqNullSafe(transcriptId)))
+  }
+
+  override def filterToExon(exonId: String): FeatureRDD = {
+    transformDataset(dataset => dataset.filter(dataset.col("exonId").eqNullSafe(exonId)))
+  }
+
+  override def filterByScore(minimumScore: Double): FeatureRDD = {
```
[the remark this answers is not in the document]
Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala]: No, I want to be careful not to go overboard and include too many methods.  I'd like to only filter out/remove numeric values less than a minimum value, consistently.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala [hunk 3]
```diff
@@ -427,6 +455,77 @@ sealed abstract class FeatureRDD extends AvroGenomicDataset[Feature, FeatureProd
...
+   *
+   * @param featureType Feature type to filter by.
+   * @return FeatureRDD filtered by the specified feature type.
+   */
+  def filterByFeatureType(featureType: String): FeatureRDD = {
+    transform(rdd => rdd.filter(f => Option(f.getFeatureType).exists(_.equals(featureType))))
+  }
+
+  /**
+   * Filter this FeatureRDD by gene.
```
[the remark this answers is not in the document]
Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala]: No, that would be `filterByFeatureType("gene")` or `filterByFeatureType("SO:0000704")`

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala]: Added in most recent commit

**Comment by fnothaft** [comment]
Merged! Thanks @heuermh!

**Pull request merged**